Lib/tarfile.py

```diff
@@ -1950,18 +1998,51 @@ def addfile(self, tarinfo, fileobj=None):
 
         tarinfo = copy.copy(tarinfo)
 
-        buf = tarinfo.tobuf(self.format, self.encoding, self.errors)
-        self.fileobj.write(buf)
-        self.offset += len(buf)
-        bufsize=self.copybufsize
+        self._write_header(tarinfo)
+        bufsize = self.copybufsize
         # If there's data to follow, append it.
         if fileobj is not None:
             copyfileobj(fileobj, self.fileobj, tarinfo.size, bufsize=bufsize)
-            blocks, remainder = divmod(tarinfo.size, BLOCKSIZE)
-            if remainder > 0:
-                self.fileobj.write(NUL * (BLOCKSIZE - remainder))
-                blocks += 1
-            self.offset += blocks * BLOCKSIZE
+            self._pad_file(tarinfo.size)
+
+        self.members.append(tarinfo)
+
+    def addbuffer(self, tarinfo, fileobj):
+        """Add the TarInfo object `tarinfo` to the archive.
+
+        This method can be used when the size of fileobj cannot be know
+        beforehand and will read from fileobj until it is exhausted.
+        Once fileobj is fully written to the archive, the size metadata will
+        be overwritten.
+
+        It is not possible to use this method when compression is used or when
+        the underlying device is not seekable.
+        """
+        self._check('awx')
+
+        if not self.fileobj.seekable():
+            raise ValueError("addbuffer can only be used on seekable media")
+
+        if self.format == PAX_FORMAT:
+            raise ValueError("addbuffer cannot be use with pax header format")
+
+        tarinfo = copy.copy(tarinfo)
+        # we record the stream as a plain file
+        tarinfo.type = REGTYPE
+        header_position = self.fileobj.tell()
+        self._write_header(tarinfo)
+
+        bufsize = self.copybufsize
+        tarinfo.size = copyfileobj(fileobj, self.fileobj, bufsize=bufsize)
```

Comment:
This is an internal copy of _tarinfo_, not the copy passed by the caller. Seems to contradict your documentation.

Reply:
This aim of the method is to write data from `fileobj` and store it as a plain file. Not overriding `tarinfo.type` cause issue when fileobj is a pipe because `tarinfo.pipe` is set as `FIFOTYPE` wich is not supposed to have data associated to it in the Tar archive as far as I can tell.

Do you think changing the signature from `addbuffer(self, tarinfo, fileobj)` to `addbuffer(self, tarinfo, fileobj, type=REGTYPE)` would be acceptable?